### ISSUE TITLE
Non-deterministic assertion violation at ../src/smt/theory_arith_inv.h Line: 198 with multi-threads

### ISSUE THREAD

**Issue by muchang** [Z3Prover/z3#3712]
Hi,
For this formula:
```
(set-option :smt.threads 10) 
(declare-fun e () Real) 
(declare-fun c () Real) 
(declare-fun a () Real) 
(declare-fun b () Real)  
(assert (> (+   b   (* (- 7) e)) 30)) 
(assert (>= (+ (* (- 3) a) (* e (- 20) e)) 0)) 
(assert (not (<= (+ (* 45 c) (mod 8 a)) 5)))    
(assert (<= (+ (* 43 c) (* 6 b) (* 4 c)) a))  
(assert (not (< (+ (* (- 34) e)) 5))) 
(assert (not (>= 0 (+  (* (- 5) c)) 1))) 
(check-sat)      
```
Z3  throws out an assertion violation:
```
[78] % ./z3/build/z3 small.smt2 
unknown
[79] % ./z3/build/z3 small.smt2
unknown
[79] % ./z3/build/z3 small.smt2
ASSERTION VIOLATION
File: ../src/smt/theory_arith_inv.h
Line: 198
!above_upper(v)
(C)ontinue, (A)bort, (S)top, (T)hrow exception, Invoke (G)DB
^[[A
INVALID COMMAND
(C)ontinue, (A)bort, (S)top, (T)hrow exception, Invoke (G)DB
INVALID COMMAND
(C)ontinue, (A)bort, (S)top, (T)hrow exception, Invoke (G)DB
[79] % ./z3/build/z3 small.smt2
unknown
[79] % ./z3/build/z3 small.smt2
unknown
[79] % ./z3/build/z3 small.smt2
unknown
[79] % ./z3/build/z3 small.smt2
unknown
[79] % ./z3/build/z3 small.smt2
unknown
[79] % ./z3/build/z3 small.smt2
unknown
[79] % ./z3/build/z3 small.smt2
unknown
[79] % ./z3/build/z3 small.smt2
unknown
[79] % ./z3/build/z3 small.smt2
unknown
[79] % ./z3/build/z3 small.smt2
unknown
[79] % ./z3/build/z3 small.smt2
unknown
[79] % ./z3/build/z3 small.smt2
unknown
[79] % ./z3/build/z3 small.smt2
unknown
[79] % ./z3/build/z3 small.smt2
ASSERTION VIOLATION
File: ../src/smt/theory_arith_inv.h
Line: 198
!above_upper(v)
(C)ontinue, (A)bort, (S)top, (T)hrow exception, Invoke (G)DB
A
[79] % cat small.smt2 
(set-option :smt.threads 10) 
(declare-fun e () Real) 
(declare-fun c () Real) 
(declare-fun a () Real) 
(declare-fun b () Real)  
(assert (> (+   b   (* (- 7) e)) 30)) 
(assert (>= (+ (* (- 3) a) (* e (- 20) e)) 0)) 
(assert (not (<= (+ (* 45 c) (mod 8 a)) 5)))    
(assert (<= (+ (* 43 c) (* 6 b) (* 4 c)) a))  
(assert (not (< (+ (* (- 34) e)) 5))) 
(assert (not (>= 0 (+  (* (- 5) c)) 1))) 
(check-sat)
```

OS: Ubuntu 18.04
Commit: c431a10

**Comment by rainoftime** [comment]
I encountered the same issue. In that test case, the violation is sensitive to the "timeout" parameter. 
The case is too long and hard to reduce (due to the non-determinism).

**Comment by zhendongsu** [comment]
See the closed https://github.com/Z3Prover/z3/issues/3522, which still triggers quite reliably.

**Comment by NikolajBjorner** [comment]
sorry, I will have to close this: it hasn't been possible to reproduce this on my side even with thousands of tries. The Windows scheduler could be different here, but I also tried my ubuntu (which could also be skewed in comparison). I updated the sanity check with a bail out if there is a cancelation: the sanity check is irrelevant under cancellation. If you get these kinds of assertions still, then a stack trace would be useful otherwise I will have to directly punt on these.
